src/tools/rbd_mirror/GroupReplayer.cc

```diff
@@ -0,0 +1,1305 @@
+
+  if (m_state == STATE_STOPPING || m_state == STATE_STOPPED) {
+    // no need to re-register if stopping
+    return;
+  }
+  locker.unlock();
+
+  unregister_admin_socket_hook();
+  register_admin_socket_hook();
+}
+
+template <typename I>
+void GroupReplayer<I>::set_mirror_group_status_update(
+    cls::rbd::MirrorGroupStatusState state, const std::string &desc) {
+  dout(20) << "state=" << state << ", description=" << desc << dendl;
+
+  cls::rbd::MirrorGroupSiteStatus local_status;
+  local_status.state = state;
+  local_status.description = desc;
+  local_status.up = true;
+
+  auto remote_status = local_status;
+
+  {
+    std::unique_lock locker{m_lock};
+    for (auto &[_, ir] : m_image_replayers) {
+      cls::rbd::MirrorImageSiteStatus mirror_image;
+      if (ir->is_running()) {
+        if (ir->is_replaying()) {
+          mirror_image.state = cls::rbd::MIRROR_IMAGE_STATUS_STATE_REPLAYING;
+        } else {
+          mirror_image.state = cls::rbd::MIRROR_IMAGE_STATUS_STATE_STARTING_REPLAY;
+        }
+      } else if (ir->is_stopped()) {
+        mirror_image.state = cls::rbd::MIRROR_IMAGE_STATUS_STATE_STOPPED;
+      } else {
+        mirror_image.state = cls::rbd::MIRROR_IMAGE_STATUS_STATE_STOPPING_REPLAY;
+      }
+      mirror_image.description = ir->get_state_description();
+
+      local_status.mirror_images[{ir->get_local_pool_id(),
+                                  ir->get_global_image_id()}] = mirror_image;
+      auto remote_pool_id = ir->get_remote_pool_id();
+      if (remote_pool_id >= 0) {
+        remote_status.mirror_images[{remote_pool_id,
+                                     ir->get_global_image_id()}] = mirror_image;
+      }
+    }
+  }
+
+  m_local_status_updater->set_mirror_group_status(m_global_group_id,
+                                                  local_status, true);
+  if (m_remote_group_peer.mirror_status_updater != nullptr) {
+    m_remote_group_peer.mirror_status_updater->set_mirror_group_status(
+        m_global_group_id, remote_status, true);
+  }
+}
+
+template <typename I>
+void GroupReplayer<I>::create_regular_group_snapshot(
+    const std::string &remote_snap_name,
+    const std::string &remote_snap_id,
+    std::vector<cls::rbd::GroupImageStatus> *local_images) {
+  // each image will have one snapshot specific to group snap, and so for each
+  // image get a ImageSnapshotSpec and prepare a vector
+  // for image :: <images in that group> {
+  //   * get snap whos name has group snap_id for that we can list snaps and
+  //     filter with remote_group_snap_id
+  //   * get its { pool_id, snap_id, image_id }
+  // }
+  // finally write to the object
+  dout(10) << dendl;
+  librados::ObjectWriteOperation op;
+  cls::rbd::GroupSnapshot group_snap{
+    remote_snap_id, // keeping it same as remote group snap id
+    cls::rbd::UserGroupSnapshotNamespace{},
+      remote_snap_name,
+      cls::rbd::GROUP_SNAPSHOT_STATE_INCOMPLETE};
+  librbd::cls_client::group_snap_set(&op, group_snap);
```

Comment:
IIUC, this doesn't actually write the snap data to the rados object. What is the requirement for doing this here?